src/test/java/com/amihaiemil/docker/AuthHttpClientTestCase.java

```diff
@@ -0,0 +1,162 @@
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.amihaiemil.docker;
+
+import java.io.IOException;
+import org.apache.http.Header;
+import org.apache.http.HttpHost;
+import org.apache.http.HttpRequest;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.ClientProtocolException;
+import org.apache.http.client.HttpClient;
+import org.apache.http.client.ResponseHandler;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.methods.HttpUriRequest;
+import org.apache.http.conn.ClientConnectionManager;
+import org.apache.http.message.BasicHeader;
+import org.apache.http.params.HttpParams;
+import org.apache.http.protocol.HttpContext;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
+import org.junit.Test;
+
+/**
+ * Unit tests for {@link AuthHttpClient}.
+ * @author George Aristy (george.aristy@gmail.com)
+ * @version $Id$
+ * @since 0.0.1
+ */
+public final class AuthHttpClientTestCase {
+    /**
+     * Must inject the X-Registry-Auth header if absent and set it to the
+     * auth's value.
+     * @throws Exception If something goes wrong.
+     */
+    @Test
+    public void injectsHeaderIfAbsent() throws Exception {
+        final HttpUriRequest request = new HttpGet();
+        new AuthHttpClient(new NoOpHttpClient(), () -> "123").execute(request);
+        MatcherAssert.assertThat(
+            request.getFirstHeader("X-Registry-Auth").getValue(),
+            Matchers.is("123")
+        );
+    }
+
+    /**
+     * Leaves the request's header instact if it exists.
+     * @throws Exception If something goes wrong.
+     */
+    @Test
+    public void leavesExistingHeaderAlone() throws Exception {
+        final Header auth = new BasicHeader("X-Registry-Auth", "12356");
+        final HttpUriRequest request = new HttpGet();
+        request.setHeader(auth);
+        new AuthHttpClient(new NoOpHttpClient(), () -> "abc").execute(request);
+        MatcherAssert.assertThat(
+            request.getFirstHeader("X-Registry-Auth"),
+            Matchers.is(auth)
+        );
+    }
+
+    /**
+     * An http client that does nothing.
+     */
+    private static class NoOpHttpClient implements HttpClient {
```

Comment:
@llorllale Do we need this class? Can't we just use ``Mockito.mock(HttpClient.class)``?

Reply:
@amihaiemil we can use Mockito, yes, although I don't like it because the mocking introduces too much noise in tests. Do you want mockito?

Reply:
@llorllale Well, yes, if the mock is not supposed to do anything, I would rather use Mockito. I understand that mocks can get really complex, but it's not the case here.